### PR TITLE
Fix autolinking errors

### LINKED ISSUE #19462: [Bug]: Auto-linking creates "pointless" `mailto:`-links in some cases
### Attach (recommended) or Link to PDF file

https://github.com/mozilla/pdf.js/blob/master/test/pdfs/pdf.pdf.link

### Web browser and its version

N/A

### Operating system and its version

N/A

### PDF.js version

`master`, or any version after PR #19110

### Is the bug present in the latest PDF.js version?

Yes

### Is a browser extension

No

### Steps to reproduce the problem

1. Load http://localhost:8888/web/viewer.html?file=/test/pdfs/pdf.pdf#page=68
2. Note that the `JD?M$0QP)lKn06l1apKDC@\qJ4B!!(5m+j.7F790m` text is clickable.

### What is the expected behavior?

Since that text doesn't make sense as an email address, there should be no link created.

*Note:* Acrobat Reader does *not* create a link for that piece of text.

### What went wrong?

The text become a `mailto:`-link, despite not making sense as an email address.

### Link to a viewer

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ryzokuken**
This commit fixes some edge cases in the autolinking logic and adds tests for them. It fixes the issues with regexp, email validation as well as link annotations being injected multiple times on zoom.

Fixes: https://github.com/mozilla/pdf.js/issues/19462

### CODE REVIEW

**Comment by calixteman** [comment]
Ideally, it'd be nice to have a separate patch with an integration test about the annotation being injected multiple times, thank you.

File: web/annotation_layer_builder.js [hunk 1]
```diff
@@ -257,6 +259,8 @@ class AnnotationLayerBuilder {
     if (!this.#externalHide) {
       this.div.hidden = false;
     }
+
+    this.#linksInjected = true;
```
Comment (Snuffleupagus) [inline web/annotation_layer_builder.js:263]: The same line is should also be added to the early-return branch above https://github.com/mozilla/pdf.js/blob/e3cca6d513d4c24393cf5d8d4af25d69b399b46e/web/annotation_layer_builder.js#L246-L248 since otherwise there'd still be unnecessary re-parsing happening (even if nothing renders in that case).

Comment (ryzokuken) [reply web/annotation_layer_builder.js:263]: Since the only other code path would be to go through with the entire rest of the function, what would be wrong with just moving the statement above this branch?

Comment (Snuffleupagus) [reply web/annotation_layer_builder.js:263]: Right, it could just be placed directly after https://github.com/mozilla/pdf.js/pull/19470/files#diff-b166eae766e0c30a0a7dba9e750bd020d5ce9b432b84d6cb722aab3a8ca95aa1R242 instead.

**Comment by ryzokuken** [comment]
@calixteman broke out the zooming fix into its own patch and added integration test.

File: test/integration/autolinker_spec.mjs [hunk 2]
```diff
@@ -121,4 +121,42 @@ describe("autolinker", function () {
...
+
+    it("must not repeatedly add link annotations redundantly", async () => {
+      await Promise.all(
+        pages.map(async ([browserName, page]) => {
+          let url = await page.$$eval(
+            ".annotationLayer > .linkAnnotation > a",
+            annotations => annotations.map(a => a.href)
+          );
+          expect(url.length).withContext(`In ${browserName}`).toEqual(1);
+          await page.evaluate(() => window.PDFViewerApplication.zoomReset());
```
Comment (Snuffleupagus) [inline test/integration/autolinker_spec.mjs:152]: I believe that there's a couple of issues with this line:
 - It *implicitly* relies on the initial zoom level of the document being "page-fit", and if the default zoom-value for the integration-tests is ever changed this test would no longer be testing the intended thing, so it'd obviously be much better to invoke zoomIn or zoomOut here.
 - Just awaiting the zoom-call probably doesn't really guarantee that re-rendering has finished *or* that the annotationLayer has been re-enabled (since it's hidden during zooming). Hence you need to wait for the appropriate events, since otherwise the checks below might not be testing the correct thing.
 - Lastly, having a blank-line before this "second part" of the test-case wouldn't hurt.

Comment (nicolo-ribaudo) [reply test/integration/autolinker_spec.mjs:152]: @ryzokuken You can take a look at this test for how to wait for the zoom to happen: https://github.com/mozilla/pdf.js/blob/d6f63d0e4b35b31355bdf4f8338e284bf72c75b2/test/integration/viewer_spec.mjs#L231

Comment (Snuffleupagus) [reply test/integration/autolinker_spec.mjs:152]: Please also see PR #19475 which adds a new event that might be helpful here, since it'll allow detecting when the inferred-links have actually been added.

Comment (ryzokuken) [reply test/integration/autolinker_spec.mjs:152]: Great, thanks for your PR, will rebase on top once it's merged and listen on that event to avoid zooming in before the links are injected.

Comment (ryzokuken) [reply test/integration/autolinker_spec.mjs]: I believe all three of your points were addressed, so I'm marking this as resolved.

**Comment by ryzokuken** [comment]
@Snuffleupagus @calixteman check out the PR now, the test waits for the link annotations and I think it should be looking much better/reliable now overall.

File: test/integration/autolinker_spec.mjs [hunk 3]
```diff
@@ -121,4 +121,47 @@ describe("autolinker", function () {
...
+
+    it("must not repeatedly add link annotations redundantly", async () => {
+      await Promise.all(
+        pages.map(async ([browserName, page]) => {
+          let url = await page.$$eval(
+            ".annotationLayer > .linkAnnotation > a",
+            annotations => annotations.map(a => a.href)
+          );
+          expect(url.length).withContext(`In ${browserName}`).toEqual(1);
+          await page.evaluate(() =>
```
Comment (nicolo-ribaudo) [inline test/integration/autolinker_spec.mjs:152]:
```suggestion

          await page.evaluate(() =>
```
(https://github.com/mozilla/pdf.js/pull/19470#discussion_r1952272752)

Comment (ryzokuken) [reply test/integration/autolinker_spec.mjs:152]: er, sorry. I made a tiny mistake pushing my changes.

**Comment by Snuffleupagus** [comment]
Taking another look at the existing test-cases, I'm not sure that any of them actually work *as intended* since they only wait for the `annotationLayer` to be available, which is however no guarantee that the inferred links have been processed.

**Comment by ryzokuken** [comment]
@Snuffleupagus fair point, although I did verify locally that the tests work as intended but I agree that it's more reliable to wait for the event to be fired before checking. Added the waits before every `$$eval` now.

File: test/integration/autolinker_spec.mjs [hunk 4]
```diff
@@ -121,4 +132,50 @@ describe("autolinker", function () {
...
+          );
+          expect(url.length).withContext(`In ${browserName}`).toEqual(1);
+
+          await page.evaluate(() =>
+            window.PDFViewerApplication.pdfViewer.updateScale({
+              drawingDelay: -1,
+              scaleFactor: 2,
+            })
+          );
+          await waitForLinkAnnotations(page);
```
Comment (nicolo-ribaudo) [inline test/integration/autolinker_spec.mjs:171]: Does this actually work? If there are no link annotations added (because we already added them), the event is not going to be fired, right?

Should we maybe add a page after the one with the links, and wait for that page after to trigger `pagerendered`? (since it will only be rendered once the first page is done rendering all the layers)

Comment (Snuffleupagus) [reply test/integration/autolinker_spec.mjs:171]: The event *should* still fire (in the same the e.g. `annotationlayerrendered`, and many others, are fired when the annotationLayer has been zoomed).

**Comment by ryzokuken** [comment]
All test failures are unrelated.

File: web/autolinker.js [hunk 5]
```diff
@@ -96,21 +96,32 @@ class Autolinker {
...
-        match[0].startsWith("mailto:") ||
         match[0].startsWith("http://") ||
         match[0].startsWith("https://")
-          ? match[0]
-          : `mailto:${match[0]}`;
+      ) {
+        raw = match[0];
+      } else {
+        // Check if the email address is valid through the URL constructor.
+        // In the future, this case be replaced by `URL.canParse`.
```
Comment (Snuffleupagus) [inline web/autolinker.js:116]: It seems that core-js has support for `URL.canParse` already, see https://github.com/zloirock/core-js?tab=readme-ov-file#url-and-urlsearchparams, so we might be able to use it directly?

File: web/autolinker.js [hunk 6]
```diff
@@ -96,21 +96,32 @@ class Autolinker {
...
+      /\b(?:https?:\/\/|mailto:|www\.)(?:[\S--[\p{P}<>]]|\/|[\S--[\[\]]]+[\S--[\p{P}<>]])+|\b[\S--[@\p{Ps}\p{Pe}<>]]+@([\S--[\p{P}<>]]+(?:\.[\S--[\p{P}<>]]+)+)/gmv;
 
     const [normalizedText, diffs] = normalize(text);
     const matches = normalizedText.matchAll(this.#regex);
     const links = [];
     for (const match of matches) {
-      const raw =
+      let raw;
+      if (
         match[0].startsWith("www.") ||
```
Comment (calixteman) [inline web/autolinker.js:109]: nit: set a variable for `match[0]`.

**Review by calixteman — APPROVED** [review]
LGTM. Thank you.

File: web/autolinker.js [hunk 7]
```diff
@@ -96,31 +96,38 @@ class Autolinker {
...
+        url.startsWith("http://") ||
+        url.startsWith("https://")
+      ) {
+        raw = url;
+      } else {
+        if (!URL.canParse(`http://${emailDomain}`)) {
+          continue;
+        }
+        raw = url.startsWith("mailto:") ? url : `mailto:${url}`;
+      }
```
Comment (Snuffleupagus) [inline web/autolinker.js:120]: Can't this be simplified/improved a little bit, now that the `try-catch` block was removed?
```suggestion
      } else if (URL.canParse(`http://${emailDomain}`)) {
        raw = url.startsWith("mailto:") ? url : `mailto:${url}`;
      } else {
        continue;
      }     
```

Comment (ryzokuken) [reply web/autolinker.js]: Fixed, thanks!
